shared/src/main/scala-2.12/magnolify/shims/package.scala

```diff
@@ -16,77 +16,15 @@
 
 package magnolify
 
-import scala.collection.generic.CanBuildFrom
-import scala.collection.mutable
-import scala.language.higherKinds
-import scala.reflect.ClassTag
+import scala.annotation.nowarn
 import scala.util.hashing.MurmurHash3
 
-package object shims {
-  trait Monadic[F[_]] extends magnolia1.Monadic[F] {
-    def flatMapS[A, B](from: F[A])(fn: A => F[B]): F[B]
-    def mapS[A, B](from: F[A])(fn: A => B): F[B]
+package object shims extends SerializableCanBuildFromInstances {
 
-    override def flatMap[A, B](from: F[A])(fn: A => F[B]): F[B] = flatMapS(from)(fn)
-    override def map[A, B](from: F[A])(fn: A => B): F[B] = mapS(from)(fn)
-  }
-
-  trait FactoryCompat[-A, +C] extends Serializable {
-    def newBuilder: mutable.Builder[A, C]
-    def build(xs: TraversableOnce[A]): C = (newBuilder ++= xs).result()
-  }
-
-  object FactoryCompat extends LowPriorityFactoryCompat1 {
-    private type FC[A, C] = FactoryCompat[A, C]
-
-    def apply[A, C](b: => mutable.Builder[A, C]): FC[A, C] =
-      new FactoryCompat[A, C] {
-        override def newBuilder: mutable.Builder[A, C] = b
-      }
-
-    implicit def arrayFC[A: ClassTag] = FactoryCompat(Array.newBuilder[A])
-    // Deprecated in 2.13
-    // implicit def traversableFC[A] = FactoryCompat(Traversable.newBuilder[A])
-    // List <: Iterable
-    // implicit def iterableFC[A] = FactoryCompat(Iterable.newBuilder[A])
-    // List <: Seq
-    // implicit def seqFC[A] = FactoryCompat(Seq.newBuilder[A])
-    // Vector <: IndexedSeq
-    // implicit def indexedSeqFC[A] = FactoryCompat(IndexedSeq.newBuilder[A])
-  }
-
-  trait LowPriorityFactoryCompat1 extends LowPriorityFactoryCompat2 {
-    implicit def listFC[A] = FactoryCompat(List.newBuilder[A])
-  }
-
-  trait LowPriorityFactoryCompat2 {
-    implicit def vectorFC[A] = FactoryCompat(Vector.newBuilder[A])
-    // Deprecated in 2.13
-    // implicit def streamFC[A] = FactoryCompat(Stream.newBuilder[A])
-  }
-
-  object SerializableCanBuildFroms {
-    private def cbf[A, C](b: => mutable.Builder[A, C]): CanBuildFrom[C, A, C] =
-      new CanBuildFrom[C, A, C] with Serializable {
-        override def apply(from: C): mutable.Builder[A, C] = b
-        override def apply(): mutable.Builder[A, C] = b
-      }
-
-    implicit def arrayCBF[A: ClassTag] = cbf(Array.newBuilder[A])
-    implicit def traversableCBF[A] = cbf(Traversable.newBuilder[A])
-    implicit def iterableCBF[A] = cbf(Iterable.newBuilder[A])
-    implicit def seqCBF[A] = cbf(Seq.newBuilder[A])
-    implicit def indexedSeqCBF[A] = cbf(IndexedSeq.newBuilder[A])
-    implicit def listCBF[A] = cbf(List.newBuilder[A])
-    implicit def vectorCBF[A] = cbf(Vector.newBuilder[A])
-    implicit def streamCBF[A] = cbf(Stream.newBuilder[A])
-  }
-
-  val JavaConverters = scala.collection.JavaConverters
-
-  def unsafeWrapArray(a: Array[Any]): Seq[Any] = mutable.WrappedArray.make(a)
+  type FactoryCompat[-A, +C] = SerializableCanBuildFrom[Nothing, A, C]
```

Comment:
`scala.collection.compat.Factory` is an alias for `CanBuildFrom` in scala 2.12. It is is not serializable, so we need our own